docs/internal_abi.md

```diff
@@ -0,0 +1,321 @@
+The compiler is free to relax any of these constraints for on-stack variables that are not passed by reference (for example, leaving some bits in a value undefined if the compiler can prove that they are never read).
+
+Verona places the following restrictions on target architectures:
+
+ - The size of a byte is 8 bits.
+ - The architecture can load and store a single byte without read-modify-write operations
+ - The order of bits within different-sized values is consistent
+ - Signed arithmetic on integers uses twos complement
+
+This is a slightly tighter set of requirements than the C specification but is looser than the de-facto C standard as used in large deployed codebases.
+
+### Booleans
+
+Boolean values are stored as bytes, with the bit pattern 0 or 1 in the platform's native endian.
+The compiler is responsible for ensuring that no bits other than the least-significant bit in a boolean are ever non-zero.
+
+### Integers
+
+Power-of-two-sized integers in Verona can be used for arithmetic and so are expected to use the representation that the target architecture uses for integer operations.
+The types `U8`, `U16`, `U32`, `U64`, and `U128 are all unsigned fixed-width integers with the platform's native endian.
+The types `S8`, `S16`, `S32`, `S64`, and `S128 are all signed fixed-width integers represented as twos complement numbers with the platform's native endian.
+These types are always stored naturally aligned.
+For example, a `U64` is always stored on an 8-byte boundary.
+
+Other integer types, for example `U56` are *storage-only* types.
+These must be extended to a power-of-two sized type for arithmetic.
+When stored on the stack, these should also be sign- or zero-extended (for signed and unsigned types, respectively) to the next native integer type.
+The value of these types is primarily for use in structures or union types, where special layout rules apply.
+
+### Floating-point values
+
+The types `Float16`, `Float32`, `Float64` and `Float128` correspond to IEEE 754 binary16, binary32, binary64, and binary128 types, respectively.
+*Note:* A future version of the language may provide IEEE decimal types, binary256, bfloat32, or smaller floating point types but these are not necessary for MVP.
+All floating-point types are stored in the target's natural floating-point byte order (which is commonly, but not always, the same as the integer byte order) and are naturally aligned.
+If the target does not support hardware floating point then floating point values are stored according to the C ABI for the platform.
+
+### Pointers
+
+We anticipate three common representations for pointers on Verona target systems:
+
+ - 32-bit integers representing an address
+ - 64-bit integers representing an address
+ - CHERI capabilities granting access to an address range.
+
+All Verona objects are at least 8-byte aligned and so the low three bits of a pointer will always be zero.
```

Comment:
What do you mean by objects here? Instances of concrete classes?

Does it mean that you can't have pointers to native types (which are naturally aligned)?

Reply:
I believe our current plan is that all pointers are pointers to objects (or sufficiently object-like things that they do have a Verona object header).